sdk/helper/testcluster/docker/environment.go

```diff
@@ -0,0 +1,967 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package docker
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/big"
+	mathrand "math/rand"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	docker "github.com/docker/docker/client"
+	"github.com/hashicorp/go-cleanhttp"
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/vault/api"
+	dockhelper "github.com/hashicorp/vault/sdk/helper/docker"
+	"github.com/hashicorp/vault/sdk/helper/logging"
+	"github.com/hashicorp/vault/sdk/helper/testcluster"
+	uberAtomic "go.uber.org/atomic"
+	"golang.org/x/net/http2"
+)
+
+var (
+	_ testcluster.VaultCluster     = &DockerCluster{}
+	_ testcluster.VaultClusterNode = &DockerClusterNode{}
+)
+
+// DockerCluster is used to managing the lifecycle of the test Vault cluster
+type DockerCluster struct {
+	ClusterName string
+
+	RaftStorage  bool
+	ClusterNodes []*DockerClusterNode
+
+	// Certificate fields
+	CACert        *x509.Certificate
+	CACertBytes   []byte
+	CACertPEM     []byte
+	CACertPEMFile string
+	CAKey         *ecdsa.PrivateKey
+	CAKeyPEM      []byte
+	RootCAs       *x509.CertPool
+
+	barrierKeys  [][]byte
+	recoveryKeys [][]byte
+	tmpDir       string
+
+	// rootToken is the initial root token created when the Vault cluster is
+	// created.
+	rootToken string
+	dockerAPI *docker.Client
+	ID        string
+	Logger    log.Logger
+}
+
+func (dc *DockerCluster) NamedLogger(s string) log.Logger {
+	return dc.Logger.Named(s)
+}
+
+func (dc *DockerCluster) ClusterID() string {
+	return dc.ID
+}
+
+func (dc *DockerCluster) Nodes() []testcluster.VaultClusterNode {
+	ret := make([]testcluster.VaultClusterNode, len(dc.ClusterNodes))
+	for i := range dc.ClusterNodes {
+		ret[i] = dc.ClusterNodes[i]
+	}
+	return ret
+}
+
+func (dc *DockerCluster) GetBarrierKeys() [][]byte {
+	return dc.barrierKeys
+}
+
+func testKeyCopy(key []byte) []byte {
+	result := make([]byte, len(key))
+	copy(result, key)
+	return result
+}
+
+func (dc *DockerCluster) GetRecoveryKeys() [][]byte {
+	ret := make([][]byte, len(dc.recoveryKeys))
+	for i, k := range dc.recoveryKeys {
+		ret[i] = testKeyCopy(k)
+	}
+	return ret
+}
+
+func (dc *DockerCluster) GetBarrierOrRecoveryKeys() [][]byte {
+	return dc.GetBarrierKeys()
```

Comment:
This comment isn't necessarily related to your changes since I know this implementation is the same as the current TestCluster, but should this function ever return recovery keys instead of barrier keys?

Reply:
Only once we support non-shamir seals.

Reply:
@ncabatoff I think it'd be really cool if we could support other base layers for the Vault container, happy to open a PR for that.

Not quite sure how it interacts with AddNodes tbh, but what I was thinking is it'd be cool to drop in a container file like:

```
FROM hashicorp/vault:latest

RUN apk install softhsm # ... plus more configuration
```

and then use PKCS#11 auto-unseal/managed keys/... testing in Vault Ent with SoftHSM. 

I think we might need some hooks to let us specify custom callbacks at points in time (startup, cluster join, &c ...) so that we can inject data from other containers (initial node will setup auto-unseal keys which we'll need to replicate into other machines unless we start a PKCS#11 network proxy client on the Vault node and transit all requests to a single, common PKCS#11 network proxy server on another node).

This is definitely enough work to be a completely separate PR though.